### PR TITLE
GODRIVER-3434 Revert builder-lister pattern for client options

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-3434

## Summary

Don't apply the builder-lister pattern for client options.

## Background & Motivation

This update has been a source of frustration for users migrating to the v2 beta. While the lister-builder pattern is effective for operational options, client options are often defined as struct literals and used in test / general comparison within a users code. For example:

```go
opts := options.Client().SetAppName("x")

// ...

if opts.AppName != nil && opts.AppName == "x" {
    // do something
}
```

While this isn't too difficult to manage in v2, it does require a code change. The primarily discussed alternative is to apply a `Merge` function to `ClientOptionsBuilder`. However, this would still require a code change and is kind of confusing for people who aren't familiar with the history of the problem in the Go Driver. This PR suggests deferring that question to v3, as client options are rather unique.

### CODE REVIEW

**Review by qingyang-hu — APPROVED** [review]
LGTM

File: mongo/options/clientoptions.go [hunk 1]
```diff
@@ -1153,38 +1013,26 @@ func (c *ClientOptionsBuilder) SetWriteConcern(wc *writeconcern.WriteConcern) *C
...
 // SetAutoEncryptionOptions specifies an AutoEncryptionOptions instance to automatically encrypt and decrypt commands
 // and their results. See the options.AutoEncryptionOptions documentation for more information about the supported
 // options.
-func (c *ClientOptionsBuilder) SetAutoEncryptionOptions(aeopts Lister[AutoEncryptionOptions]) *ClientOptionsBuilder {
-	c.Opts = append(c.Opts, func(opts *ClientOptions) error {
-		opts.AutoEncryptionOptions = aeopts
-
-		return nil
-	})
+func (c *ClientOptions) SetAutoEncryptionOptions(aeopts Lister[AutoEncryptionOptions]) *ClientOptions {
```
Comment (matthewdale) [inline mongo/options/clientoptions.go:1034]: For `AutoEncryptionOptions`, `LoggerOptions`, and `ServerAPIOptions`, should we also revert to the previous pattern (i.e. not use of `Lister` type)?

**Review by matthewdale — APPROVED** [review]
Looks good! 👍

File: mongo/options/serverapioptions.go [hunk 2]
```diff
@@ -29,50 +29,22 @@ type ServerAPIOptions struct {
...
-		return nil
-	})
-
-	return opts
-}
-
-// List returns a list of ServerAPIOptions setter functions.
-func (s *ServerAPIOptionsBuilder) List() []func(*ServerAPIOptions) error {
-	return s.Opts
+// serverAPIversion.// ServerAPI creates a new ServerAPIOptions configured with the provided serverAPIversion.
```
Comment (qingyang-hu) [inline mongo/options/serverapioptions.go:33]:
``` suggestion
// serverAPIversion.
```

File: mongo/client.go [hunk 3]
```diff
@@ -1,5 +1,5 @@
-// Copyright (C) MongoDB, Inc. 2017-present.
 //
+// Copyright (C) MongoDB, Inc. 2017-present.
```
Comment (qingyang-hu) [inline mongo/client.go:2]:
``` suggestion
// Copyright (C) MongoDB, Inc. 2017-present.
//
```

**Review by qingyang-hu — CHANGES_REQUESTED** [review]
It looks good other than some typos.

**Review by qingyang-hu — APPROVED** [review]
👍

**Review by matthewdale — APPROVED** [review]

**Pull request merged**